### PR TITLE
Eliminate the needs to start a Pyspark session to use SQL ML

### PR DESCRIPTION
**Pull request by eddyxu**
To make Rikai more suitable for a service, this PR  re-architecture the way that Spark SQL engine manages ML model code from Python.

We can start a Spark SQL with `ML_Predict` without starting `pyspark`, as long as we have the configuration in place (i.e., in `spark-defaults.conf`) and set propriety Python versions in driver and worker nodes.  

```
spark.sql.extensions ai.eto.rikai.sql.spark.RikaiSparkSessionExtensions
```

Spark can pick up the right udf to run python models, like pytorch, tf and sklearn. It could open more use cases / deployment modes for Rikai, i.e., can use Rikai with Livy, or Spark thrift server then

### CODE REVIEW

**Comment by eddyxu** [comment]
Update: Only works for FileSystemRegistry model spec, will follow up with MLflowRegistry and others later in the following days.

File: src/main/scala/org/apache/spark/sql/rikai/model/Resolver.scala [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                 |    fobj.write(pickle.dumps((func, dataType)))
+                 |with open("${dataTypePath}", "w") as fobj:
+                 |    fobj.write(dataType.json())
+                 |""".stripMargin)
+      val cmd = Files.readAllBytes(path)
+      val dataTypeJson = Files.readString(dataTypePath)
+      val returnType = DataType.fromJson(dataTypeJson)
+      val udf =
+        UserDefinedPythonFunction(
+          "test_sum",
```
Comment (eddyxu) [inline src/main/scala/org/apache/spark/sql/rikai/model/Resolver.scala:59]: ignore these naming details for now. The current state proves that the mechanism works

**Comment by eddyxu** [comment]
Ran `sbt "testOnly **.MLPredictTest"`

```
+--------------------+
|                   s|
+--------------------+
|{[[3.70917, 230.8...|
+--------------------+
root
 |-- s: struct (nullable = true)
 |    |-- boxes: array (nullable = true)
 |    |    |-- element: array (containsNull = true)
 |    |    |    |-- element: float (containsNull = true)
 |    |-- scores: array (nullable = true)
 |    |    |-- element: float (containsNull = true)
 |    |-- label_ids: array (nullable = true)
 |    |    |-- element: integer (containsNull = true)
```

File: src/main/scala/org/apache/spark/sql/rikai/model/Resolver.scala [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+                 |""".stripMargin,
+        session
+      )
+      val cmd = Files.readAllBytes(path)
+      val dataTypeJson = Files.readString(dataTypePath)
+      val returnType = DataType.fromJson(dataTypeJson)
+      val suffix = Random.alphanumeric.take(6)
+      val udfName = s"${spec.name}_${suffix}"
+      val udf =
+        UserDefinedPythonFunction(
```
Comment (changhiskhan) [inline src/main/scala/org/apache/spark/sql/rikai/model/Resolver.scala:64]: can we also use this mechanism to expose the other rikai pyspark udfs? (doesn't have to be done in this PR)

File: src/main/scala/org/apache/spark/sql/rikai/model/Resolver.scala [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      session: SparkSession,
+      registryClassName: String,
+      spec: ModelSpec
+  ): SparkUDFModel = {
+    val specPath = Files.createTempFile("model-spec", ".json")
+    val path = Files.createTempFile("model-code", ".cpt")
+    val dataTypePath = Files.createTempFile("model-type", ".json")
+    try {
+      Files.writeString(specPath, spec.asJson.toString)
+      Python.execute(
```
Comment (changhiskhan) [inline src/main/scala/org/apache/spark/sql/rikai/model/Resolver.scala:44]: what does the error look like here if say the spec is bad?

File: src/main/scala/org/apache/spark/sql/rikai/model/Resolver.scala [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            Python.pythonExec,
+            Python.pythonVer,
+            Seq.empty.asJava,
+            null
+          ),
+          returnType,
+          PythonEvalType.SQL_SCALAR_PANDAS_ITER_UDF,
+          udfDeterministic = true
+        )
+      session.udf.registerPython(udfName, udf)
```
Comment (changhiskhan) [inline src/main/scala/org/apache/spark/sql/rikai/model/Resolver.scala:79]: is this being done on every invocation? or just the first time when this udf is created?

**Review by changhiskhan — COMMENTED** [review]
very exciting! Some questions for my own clarification

File: src/main/scala/org/apache/spark/sql/rikai/model/Resolver.scala (hunk 2 above)
Comment (eddyxu) [reply src/main/scala/org/apache/spark/sql/rikai/model/Resolver.scala:64]: yes, so basically just use the same python interpreter to pickle function and its return `DataType`, you can register any functions (with some size limitations).

This is how pyspark registers udfs and ships them to workers.

File: src/main/scala/org/apache/spark/sql/rikai/model/Resolver.scala (hunk 3 above)
Comment (eddyxu) [reply src/main/scala/org/apache/spark/sql/rikai/model/Resolver.scala:44]: It throws a `RuntimeError(captured_stderr)`

File: src/main/scala/org/apache/spark/sql/rikai/model/Resolver.scala (hunk 4 above)
Comment (eddyxu) [reply src/main/scala/org/apache/spark/sql/rikai/model/Resolver.scala:79]: this is everytime a model / udf is created / registered.

File: python/rikai/spark/sql/codegen/mlflow_registry.py [hunk 5]
```diff
@@ -187,11 +183,11 @@ def tracking_client(self):
 
     @property
     def mlflow_tracking_uri(self):
-        return self._spark.conf.get(CONF_MLFLOW_TRACKING_URI)
+        return os.environ.get(CONF_MLFLOW_TRACKING_URI)
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/rikai/spark/sql/codegen/mlflow_registry.py:186]: So the JIT codegen does not have SparkSession, it is just a pure python process to pickle python code.

File: python/rikai/spark/sql/codegen/base.py [hunk 6]
```diff
@@ -201,34 +201,8 @@ def udf_from_spec(spec: ModelSpec):
...
-    Returns
-    -------
-    str
-        Spark UDF function name for the generated data.
-    """
-    udf = udf_from_spec(spec)
-    return register_udf(spark, udf, name)
+def command_from_spec(registry_class: str, row_spec: dict):
+    cls = find_class(registry_class)
+    resolver = cls()
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/rikai/spark/sql/codegen/base.py:206]: lemme rename this to `registry`

File: src/main/scala/ai/eto/rikai/sql/spark/Python.scala [hunk 7]
```diff
@@ -16,52 +16,47 @@
...
+import scala.sys.process
+import scala.sys.process.{Process, ProcessLogger}
 
 object Python {
-  private var python: Option[Python] = None
 
-  def register(mr: Python): Unit =
-    python = Some(mr)
+  /** Python executor */
+  lazy val pythonExec: String =
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply src/main/scala/ai/eto/rikai/sql/spark/Python.scala:27]: None that is obvious to me. 

Guess that was the name that Spark developers made something similar in their tests

https://github.com/apache/spark/blob/11596b3b17b5e0f54e104cd49b1397c33c34719d/sql/core/src/test/scala/org/apache/spark/sql/IntegratedUDFTestUtils.scala#L197

**Review by changhiskhan — APPROVED** [review]
This is awesome

**Pull request merged**